[Merged by Bors] - feat: the completion of a number field at an infinite place

smmercuri opened the pull request:
This PR contains the following:

- The completion `NumberField.InfinitePlace.completion` of a number field at an infinite place, which is derived from `AbsoluteValue.completion` as `InfinitePlace` is a subtype of `AbsoluteValue`.
- Proofs that if an infinite place of a number field is real/complex then the resultant completion is isomorphic to the real/complex numbers.

cf. #13577 for previous feedback

---

- [x] depends on: #17998
- [x] depends on: #18144 

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

riccardobrasca commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,344 @@
+
+/-- Type synonym equipping a semiring with an absolute value. -/
+@[nolint unusedArguments]
+def WithAbs {R S} [Semiring R] [OrderedSemiring S] : AbsoluteValue R S → Type _ := fun _ => R
```
```suggestion
def WithAbs {R S : Type*} [Semiring R] [OrderedSemiring S] : AbsoluteValue R S → Type _ := fun _ => R
```
And similarly elsewhere.

smmercuri replied:
Done!

riccardobrasca commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,344 @@
+
+namespace WithAbs
+
+instance {R} [Ring R] (v : AbsoluteValue R ℝ) : NormedRing (WithAbs v) where
```
Why not making this a `def`, without using `WithAbs`? (It should be in some other file I think).

smmercuri replied:
Do you mean doing something like the following? 

```
def normedRing {R : Type*} [Ring R] (v : AbsoluteValue R ℝ) : NormedRing R where 
  dist_eq _ _ := rfl
  dist_self x := by simp only [sub_self, MulHom.toFun_eq_coe, AbsoluteValue.coe_toMulHom, map_zero]
  dist_comm := v.map_sub
  dist_triangle := v.sub_le
  norm_mul x y := (v.map_mul x y).le
  eq_of_dist_eq_zero := by simp only [MulHom.toFun_eq_coe, AbsoluteValue.coe_toMulHom,
    AbsoluteValue.map_sub_eq_zero_iff, imp_self, implies_true]

instance {R : Type*} [Ring R] (v : AbsoluteValue R ℝ) : NormedRing (WithAbs v) := 
  letI : Ring (WithAbs v) := inferInstanceAs (Ring R)
  normedRing v
```

and placing the `normedRing` `def` in an `AbsoluteValue` related file?

riccardobrasca replied:
Yes, you can call this something like `AbsoluteValue.normedRing` so people can write `v.normedRing`. But it is just an idea, I am not sure it will make things easier.

riccardobrasca replied:
It seems to me that the situation is analogous to [Valuation](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Valuation/Basic.html#Valuation)[Valuation](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Valuation/Basic.html#Valuation) and [Valued](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued)[Valued](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued). It seems that mathlib's strategy is to use `Valuation.mk` to turn a valuation into an instance of `Valued` (something you can do locally in a file) and then play with it. I think that here the situation is the same, with a valuation replaced by an absolute value and `Valued` replaced by `NormedRing`.

smmercuri replied:
Yes, this situation is analogous to the non-Archimedean case. Actually, in my original approach in #13577  I used the model of [`Valued`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued) and [`HeightOneSpectrum.adicCompletion`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/DedekindDomain/AdicValuation.html#IsDedekindDomain.HeightOneSpectrum.adicCompletion). This worked fine, but then I needed to use either lots of `@`s or `letI := v.normedField`s everywhere to explicitly tell the system which uniform space structure I needed. The type synonym approach of `WithAbs` lets the inference system take care of this, so it seems a bit neater to me, but we can keep `WithAbs` and its instances while abstracting the `normedRing` definitions to an `AbsoluteValue` file/namespace as you suggest. I will take a look!

faenuccio replied:
Oh, this is strange and it might showcase some deficiency in the `Valued` API. Have you tried there to define a `local instance` of uniform space using `v.normedField`? I do have the feeling that having both valuations and absolute values (that are the same gadget, after all) is a problem for the library. Or do you find any advantage of using the latter rather than the former (beyond the problem that `InfinitePlaces` are defined using `AbsoluteValue` but this can be refactored)?

faenuccio replied:
Yes, that's true.

riccardobrasca commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,344 @@
+induced by `f`. -/
+theorem pseudoMetricSpace_induced_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    (instNormedFieldWithAbs v).toPseudoMetricSpace = PseudoMetricSpace.induced f inferInstance := by
```
It is probably a good idea to give a more useful name to `instNormedFieldWithAbs` if you want to use it.

riccardobrasca commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,344 @@
+variable {K : Type*} [Field K] (v : AbsoluteValue K ℝ)
+
+/-- The completion of a field with respect to a real absolute value. -/
+def completion := UniformSpace.Completion (WithAbs v)
```
It is maybe better to make it an `abbrev` or something.

smmercuri commented:
> Have you discussed `WithAbs` on Zulip? It seems a reasonable idea, but maybe we can came up with something better.

There was some discussion about introducing `WithAbs` on my older PR #13577 (which did not have `WithAbs` to begin with, and its introduction helped a lot). Do you think this is worth discussing further on Zulip?

faenuccio commented:
> > Have you discussed `WithAbs` on Zulip? It seems a reasonable idea, but maybe we can came up with something better.
> 
> There was some discussion about introducing `WithAbs` on my older PR #13577 (which did not have `WithAbs` to begin with, and its introduction helped a lot). Do you think this is worth discussing further on Zulip?

Well, perhaps can you at least explain here and in the doc of the file what the main idea is? Can't you simply make do with the `AbsoluteValue R $$\mathbb{R}$$` assumption?

smmercuri commented:
> > > Have you discussed `WithAbs` on Zulip? It seems a reasonable idea, but maybe we can came up with something better.
> > 
> > 
> > There was some discussion about introducing `WithAbs` on my older PR #13577 (which did not have `WithAbs` to begin with, and its introduction helped a lot). Do you think this is worth discussing further on Zulip?
> 
> Well, perhaps can you at least explain here and in the doc of the file what the main idea is? Can't you simply make do with the `AbsoluteValue R $\mathbb{R}$` assumption?

Yes let me update the description for this PR and add more detail in the file header. It is possible to do this without `WithAbs` and using only `(v : AbsoluteValue R ℝ)`. Then we might have something like `def AbsoluteValue.normedField : NormedField K ...`. This is similar to my original approach in #13577. In that case then we either need to use lots of `letI := v.normedField` or `@`s to tell the inference system which uniform space instances we mean (at least for the `AbsoluteValue` completion part). The type synonym approach of `WithAbs` allows the inference system take care of this, so it seems a bit neater to me,  although I'm happy to discuss further!

faenuccio commented:
> > > > Have you discussed `WithAbs` on Zulip? It seems a reasonable idea, but maybe we can came up with something better.
> > > 
> > > 
> > > There was some discussion about introducing `WithAbs` on my older PR #13577 (which did not have `WithAbs` to begin with, and its introduction helped a lot). Do you think this is worth discussing further on Zulip?
> > 
> > 
> > Well, perhaps can you at least explain here and in the doc of the file what the main idea is? Can't you simply make do with the `AbsoluteValue R $\mathbb{R}$` assumption?
> 
> Yes let me update the description for this PR and add more detail in the file header. It is possible to do this without `WithAbs` and using only `(v : AbsoluteValue R ℝ)`. Then we might have something like `def AbsoluteValue.normedField : NormedField K ...`. This is similar to my original approach in #13577. In that case then we either need to use lots of `letI := v.normedField` or `@`s to tell the inference system which uniform space instances we mean (at least for the `AbsoluteValue` completion part). The type synonym approach of `WithAbs` allows the inference system take care of this, so it seems a bit neater to me, although I'm happy to discuss further!


Well, we certainly prefer neat code ;-) So I'd like to understand better if your `@` business can be avoided by some fine use of the tools at our disposal, of there is really something deep going on. I do not have anything agains your `WithAbs` approach, but if a unified one can be used for both the archimedean and the non-archimedean situation, that would be preferable.

faenuccio commented:
> > Have you discussed `WithAbs` on Zulip? It seems a reasonable idea, but maybe we can came up with something better.
> 
> There was some discussion about introducing `WithAbs` on my older PR #13577 (which did not have `WithAbs` to begin with, and its introduction helped a lot). Do you think this is worth discussing further on Zulip?

BTW Can you link to the specific point of the long thread of discussions in that PR where this has been discussed? Just to avoid you repeating old arguments...

smmercuri commented:
> > > > > Have you discussed `WithAbs` on Zulip? It seems a reasonable idea, but maybe we can came up with something better.
> > > > 
> > > > 
> > > > There was some discussion about introducing `WithAbs` on my older PR #13577 (which did not have `WithAbs` to begin with, and its introduction helped a lot). Do you think this is worth discussing further on Zulip?
> > > 
> > > 
> > > Well, perhaps can you at least explain here and in the doc of the file what the main idea is? Can't you simply make do with the `AbsoluteValue R $\mathbb{R}$` assumption?
> > 
> > 
> > Yes let me update the description for this PR and add more detail in the file header. It is possible to do this without `WithAbs` and using only `(v : AbsoluteValue R ℝ)`. Then we might have something like `def AbsoluteValue.normedField : NormedField K ...`. This is similar to my original approach in #13577. In that case then we either need to use lots of `letI := v.normedField` or `@`s to tell the inference system which uniform space instances we mean (at least for the `AbsoluteValue` completion part). The type synonym approach of `WithAbs` allows the inference system take care of this, so it seems a bit neater to me, although I'm happy to discuss further!
> 
> Well, we certainly prefer neat code ;-) So I'd like to understand better if your `@` business can be avoided by some fine use of the tools at our disposal, of there is really something deep going on. I do not have anything agains your `WithAbs` approach, but if a unified one can be used for both the archimedean and the non-archimedean situation, that would be preferable.

It seems to me that it helps to think about `v`-adic valuations as discrete integer-valued multiplicative valuations, instead of real absolute values, since e.g. then we don't need to make a choice of base (probably there are better reasons!). But, since this is not possible for the infinite places, to fully unify them we could turn a `v`-adic valuation into a `v`-adic absolute value, then we might use `AbsoluteValue.completion` for both the finite and infinite places. But that sounds like a lot of refactoring for the `adicCompletion` API, and I imagine there's a lot of material from the `Valued` API that is made use of for `adicCompletion` and is not available for `AbsoluteValue`.

smmercuri commented:
> > > Have you discussed `WithAbs` on Zulip? It seems a reasonable idea, but maybe we can came up with something better.
> > 
> > 
> > There was some discussion about introducing `WithAbs` on my older PR #13577 (which did not have `WithAbs` to begin with, and its introduction helped a lot). Do you think this is worth discussing further on Zulip?
> 
> BTW Can you link to the specific point of the long thread of discussions in that PR where this has been discussed? Just to avoid you repeating old arguments...

Oh yes sorry, I think around [Andrew's suggestion](https://github.com/leanprover-community/mathlib4/pull/13577#issuecomment-2158320382) although looking back it seems like the main point of agreement was abstracting to `AbsoluteValue` rather than the use of `WithAbs` itself.

faenuccio commented:
> > > > > Have you discussed `WithAbs` on Zulip? It seems a reasonable idea, but maybe we can came up with something better.
> > > > 
> > > > 
> > > > There was some discussion about introducing `WithAbs` on my older PR #13577 (which did not have `WithAbs` to begin with, and its introduction helped a lot). Do you think this is worth discussing further on Zulip?
> > > 
> > > 
> > > Well, perhaps can you at least explain here and in the doc of the file what the main idea is? Can't you simply make do with the `AbsoluteValue R $\mathbb{R}$` assumption?
> > 
> > 
> > Yes let me update the description for this PR and add more detail in the file header. It is possible to do this without `WithAbs` and using only `(v : AbsoluteValue R ℝ)`. Then we might have something like `def AbsoluteValue.normedField : NormedField K ...`. This is similar to my original approach in #13577. In that case then we either need to use lots of `letI := v.normedField` or `@`s to tell the inference system which uniform space instances we mean (at least for the `AbsoluteValue` completion part). The type synonym approach of `WithAbs` allows the inference system take care of this, so it seems a bit neater to me, although I'm happy to discuss further!
> 
> Well, we certainly prefer neat code ;-) So I'd like to understand better if your `@` business can be avoided by some fine use of the tools at our disposal, of there is really something deep going on. I do not have anything agains your `WithAbs` approach, but if a unified one can be used for both the archimedean and the non-archimedean situation, that would be preferable.

Oh no, you're right. Forget my suggestion to use `Valuation`. But I guess that the main point we were trying to make with @riccardobrasca is that in the same vein as [Valued.mk'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued.mk') associates a valued structure to a Valuation, it seems that simply creating a `NormedRing.mk` that takes an absolute value and creates a normed ring might be a good solution. At least, it works very well in the nonarchimedean setting, _mutatis mutandis._ The rest of your PR would be hardly modified, I guess.

smmercuri commented:
I think I understand! In the latest commit I have done the following:
- Defined `def normedField` inside the `AbsoluteValue` namespace in a separate file that takes an `AbsoluteValue` and returns a `NormedField` class -- I think this is neater and makes the `NormedField` construction available for other usage, so thanks for the suggestion @riccardobrasca !
- Removed the `NormedRing` and `NormOneClass` instances. I did not replace these with `def` versions in the `AbsoluteValue` namespace, because I realised I don't actually need them. Let me know if you think I should add them anyway for completeness -- I guess `NormedRing` is the lowest possible level for `AbsoluteValue`.
- I have kept the usage of `WithAbs` and included a paragraph in the documentation to explain a bit about this approach. If we remove it then we will need to use `@` (like the [non-Archimedean case](https://github.com/leanprover-community/mathlib4/blob/6dcd0d943385f315b956fc76968c842a3ad4072c/Mathlib/RingTheory/DedekindDomain/AdicValuation.lean#L365)) or `letI := ...`, which I had in the old PR. I do not know personally which is preferred, but the type synonym leads to shorter constructions I guess!
- Also updated the PR description.

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+  when `v` is real.
+
+## Tags
+number field, embeddings, infinite places, completion
```
```suggestion
number field, embeddings, infinite places, completion, absolute value
```

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+-/
+noncomputable section
+
+/-- Type synonym equipping a semiring with an absolute value. -/
```
Once you'll have expanded the docstring, you can copy-paste here the relevant explanation/description of this construction.

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+    ← f.map_sub, h]
+  rfl
+
+instance : Inhabited (WithAbs v) := ⟨0⟩
```
Do you need it here? It seems to me that you just use it below when speaking about completion.

smmercuri replied:
Not necessarily, but I think it should be in the `WithAbs` namespace. Have moved it to the top.

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+/-- If the absolute value `v` factors through an embedding `f` into a normed field, then
+the pseudo metric space associated to the absolute value is the same as the pseudo metric space
+induced by `f`. -/
+theorem pseudoMetricSpace_induced_of_comp
```
In the four declarations below, I would be tempted to replace `..._of_comp` by `..._of_isComp`; the assumption is that `v` is a composition, not that if you compose it with a function then something happens. Can you check similar other examples in the library?

smmercuri replied:
I think `isComp` should only be used if there is an `IsComp` structure/class that we are directly using. We could use `..._of_is_comp`, but it seems like other examples use just `_of_comp` -- see [`LinearIndependent.of_comp`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/LinearIndependent.html#LinearIndependent.of_comp)

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+
+variable {K : Type*} [Field K] (v : AbsoluteValue K ℝ)
+
+instance normedField : NormedField (WithAbs v) :=
```
If you make it for a `NormedRing`, is there a chance that in the case of fields everything goes smoothly or do you need to re-do everything? In the first case, it is worth generalising it, but in the second case of course no, leave it as it is.

smmercuri replied:
This is possible, but it will require a bit of extra work. Further down in the file I want to show that the completion of a number field at an infinite place is indeed a field, so I'm using [this](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/UniformField.html#UniformSpace.Completion.instField), which requires a `TopologicalDivisionRing` instance on `K`. This comes for free if `K` is a `NormedField` but not if `K` is a `NormedRing`. On the other hand, for the non-Archimedean case, `Valued` class carries a [`TopologicalDivisionRing`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuedField.html#Valued.topologicalDivisionRing) instance so it all works smoothly using `Valued` there.

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+
+namespace Completion
+
+instance : NormedRing v.completion :=
```
Why do you spell the instances below as such? They can automatically be inferred by type-class inference (if you replace the proof by `inferInstance`, lean is happy).

smmercuri replied:
Yes these can mostly be removed, except `Coe`.

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+variable {L : Type*} [NormedField L] [CompleteSpace L] {f : WithAbs v →+* L} {v}
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field
+`A`, then we can extend that embedding to an embedding on the completion `v.completion →+* A`. -/
```
```suggestion
`L`, then we can extend that embedding to an embedding on the completion `v.completion →+* L`. -/
```

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+  rfl
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* A` preserves distances. -/
```
```suggestion
then the extended embedding `v.completion →+* L` preserves distances. -/
```

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+    ClosedEmbedding (extensionEmbedding_of_comp h) :=
+  (isometry_extensionEmbedding_of_comp h).closedEmbedding
+
+/-- If the absolute value of a normed field factors through an embedding into another normed and
+locally compact field, then the completion of the first normed field is also locally compact. -/
```
```suggestion
/-- If the absolute value of a normed field factors through an embedding into another normed field 
that is locally compact, then the completion of the first normed field is also locally compact. -/
```

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+
+open AbsoluteValue.Completion
+
+variable {K : Type*} [Field K] (v : InfinitePlace K)
```
I believe that these result really belong to the main file concerning infinite places, they do not have much to do with completions, no?

smmercuri replied:
Agreed, I moved them.

faenuccio commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+/-- The completion of a number field at an infinite place. -/
+abbrev completion := v.1.completion
+
+namespace Completion
```
Can you add a couple of lines as "Implementation Details" in the docstring, saying why you need to re-do the work passing from Absolute Values to Infinite Places? Have you tried to define a coercion instance from the second to the first to see if this makes it possible to write replace all the `v.1`'s simply by `v` (for an InfinitePlace `v`), so that the instances below are found automatically? A wonderful example of a file with a nice "Implementation Detail" section is Mathlib.MeasureTheory.Function.LpSpace.

smmercuri replied:
Actually, the instances are already found automatically (except for `Algebra`) so I have removed them. Defining a coercion from `InfinitePlace` to `AbsoluteValue` does not seem to make much difference, and I think the duplication is mostly now gone. Let me know if you still think an "Implementation Detail" section would be helpful (perhaps the `WithAbs` explanation could go there too if so)

faenuccio reviewed (commented):
This is an excellent PR, and I am more and more in favour of adopting it. Nevertheless, I would suggest that the general description about `WithAbs` is moved in the general file about normed fields (in `Analysis/Normed/Field/Basic` or whereabouts) and only results pertaining to number fields (and infinite places) stay here.

I am stopping halfway of the `Completion`  file (having checked all the others), right at the beginning of the `NumberField.InfinitePlace.Completion` namespace.

faenuccio commented:
On top of all my previous comments, I have been thinking that perhaps another solution instead of your type alias would be to define a structure (probably not upgraded to a class, but I would expand on this at a later stage) `WithAbsRing` with two fields: a ring and an absolute value on it. Have you tried this idea? I think it would make your life easier both when trying to construct a sub-WithAbsRing from a given one; and when constructing a `WithAbsRing` from an infinite place, by doing `WithAbsRing.mk K v.1` (for `K` the field and `v` the infinite place).

Perhaps a word about why "type synonyms" are raising so much scepticism: there is nothing inherently bad with them, and there are very high chances that your solution is excellent. Nevertheless, they are seldom used in the library, whereas bundling things through `structures` appears everywhere. I have the feeling that your guiding principle here is not really "create another copy of `R` that I want to call something else than `Ring` for whatever reason"; rather, to bundle together the underlying ring and an absolute value on it, and we're used to do so with structures. Clearly, "we're used to..." is a social thing, and I would be very happy to be convinced that in this case your suggestion is far better; on the other hand, surfing the wave and benefitting of all the infrastructure built around structures (pun not intended) could be useful.

Any thoughts on this @riccardobrasca ?

riccardobrasca commented:
> On top of all my previous comments, I have been thinking that perhaps another solution instead of your type alias would be to define a structure (probably not upgraded to a class, but I would expand on this at a later stage) `WithAbsRing` with two fields: a ring and an absolute value on it. Have you tried this idea? I think it would make your life easier both when trying to construct a sub-WithAbsRing from a given one; and when constructing a `WithAbsRing` from an infinite place, by doing `WithAbsRing.mk K v.1` (for `K` the field and `v` the infinite place). All your duplicated code with the `inferInstanceAs` would be automatically OK, once you prove your results (like those about an absolute value that factors through a complex embedding), and attach to an infinite place a term of this structure as above.
> 
> Perhaps a word about why "type synonyms" are raising so much scepticism: there is nothing inherently bad with them, and there are very high chances that your solution is excellent. Nevertheless, they are seldom used in the library, whereas bundling things through `structures` appears everywhere. I have the feeling that your guiding principle here is not really "create another copy of `R` that I want to call something else than `Ring` for whatever reason"; rather, to bundle together the underlying ring and an absolute value on it, and we're used to do so with structures. Clearly, "we're used to..." is a social thing, and I would be very happy to be convinced that in this case your suggestion is far better; on the other hand, surfing the wave and benefitting of all the infrastructure built around structures (pun not intended) could be useful.
> 
> Any thoughts on this @riccardobrasca ?

I think that in Lean4 we prefer structure (even with only one field given by a term of type `R`) to avoid type leakage (meaning that sometimes Lean still see instances of the first type). Anyway my concern was mainly about not having a different solution for valuations and absolute values.

smmercuri commented:
> On top of all my previous comments, I have been thinking that perhaps another solution instead of your type alias would be to define a structure (probably not upgraded to a class, but I would expand on this at a later stage) `WithAbsRing` with two fields: a ring and an absolute value on it. Have you tried this idea? I think it would make your life easier both when trying to construct a sub-WithAbsRing from a given one; and when constructing a `WithAbsRing` from an infinite place, by doing `WithAbsRing.mk K v.1` (for `K` the field and `v` the infinite place). All your duplicated code with the `inferInstanceAs` would be automatically OK, once you prove your results (like those about an absolute value that factors through a complex embedding), and attach to an infinite place a term of this structure as above.
> 
> Perhaps a word about why "type synonyms" are raising so much scepticism: there is nothing inherently bad with them, and there are very high chances that your solution is excellent. Nevertheless, they are seldom used in the library, whereas bundling things through `structures` appears everywhere. I have the feeling that your guiding principle here is not really "create another copy of `R` that I want to call something else than `Ring` for whatever reason"; rather, to bundle together the underlying ring and an absolute value on it, and we're used to do so with structures. Clearly, "we're used to..." is a social thing, and I would be very happy to be convinced that in this case your suggestion is far better; on the other hand, surfing the wave and benefitting of all the infrastructure built around structures (pun not intended) could be useful.
> 
> Any thoughts on this @riccardobrasca ?

Thanks for the review! Yes I think the structure approach would also work, so I will give it a try

smmercuri commented:
> On top of all my previous comments, I have been thinking that perhaps another solution instead of your type alias would be to define a structure (probably not upgraded to a class, but I would expand on this at a later stage) `WithAbsRing` with two fields: a ring and an absolute value on it. Have you tried this idea? I think it would make your life easier both when trying to construct a sub-WithAbsRing from a given one; and when constructing a `WithAbsRing` from an infinite place, by doing `WithAbsRing.mk K v.1` (for `K` the field and `v` the infinite place). All your duplicated code with the `inferInstanceAs` would be automatically OK, once you prove your results (like those about an absolute value that factors through a complex embedding), and attach to an infinite place a term of this structure as above.
> 

I'm not sure if this is what you had in mind, but I tried this as the following
```
structure AbsoluteValued (R S : Type*) [Semiring R] [OrderedSemiring S]
    where
  abs : AbsoluteValue R S
  value : R
```
But it's not clear to me where to go from here with such a structure. I think `class` might be better because it should extend `Semiring` and `NormedField`, so that it has access to these instances I think? I have not interacted much with `structure` so this is probably my ignorance at play here. 

I think it might be helpful to make a more concrete analogy to the non-Archimedean case to clarify the options here. In the non-Archimedean case we want to use the [`Valued`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued) class which carries the `UniformSpace` structure we are interested in, but this has a distinguished valuation whereas we have multiple valuations. So there is a constructor [`Valued.mk'`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued.mk') which constructs the `Valued` instance when there is no preferred valuation/uniform structure. Then a `def` is made for [`adicValued`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/DedekindDomain/AdicValuation.html#IsDedekindDomain.HeightOneSpectrum.adicValued) which carries these `Valued` instances. In particular, [`adicCompletion`](https://github.com/leanprover-community/mathlib4/blob/3ca1060d7e48260d1b1b7a0733f887e455455226/Mathlib/RingTheory/DedekindDomain/AdicValuation.lean#L365-L366) calls `adicValued.toUniformSpace` using the `@` operator.

Now for the Archimedean case, the [`NormedField`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Field/Basic.html#NormedField) class plays the role of `Valued`. Again this has a distinguished norm, so we want a way of creating an instance for non-distinguished norms (or absolute values in this case). I think there are two main options here: (1) in analogy to non-Arch. case we create a `NormedField.mk'` constructor taking an absolute value and giving a `NormedField` instance, assign this to a `def`, then use `@` operators as appropriate; (2) create a type synonym that depends on absolute values, in order to assign the `NormedField` instance as in this PR. It's not clear to me where a new structure fits in to these options. 

By the way, I did not realise this but those `inferInstanceAs` can actually just be deleted with the type synonym approach as well (except for `Coe` I think).

> Perhaps a word about why "type synonyms" are raising so much scepticism: there is nothing inherently bad with them, and there are very high chances that your solution is excellent. Nevertheless, they are seldom used in the library, whereas bundling things through `structures` appears everywhere. I have the feeling that your guiding principle here is not really "create another copy of `R` that I want to call something else than `Ring` for whatever reason"; rather, to bundle together the underlying ring and an absolute value on it, and we're used to do so with structures. Clearly, "we're used to..." is a social thing, and I would be very happy to be convinced that in this case your suggestion is far better; on the other hand, surfing the wave and benefitting of all the infrastructure built around structures (pun not intended) could be useful.
> 
I agree with a previous comment that the use of "equip" is misleading in the documentation, and I can see where the scepticism is coming from! My guiding principle for using a type synonym is less about "equipping" the ring with an absolute value, but really more about providing a mechanism for type class inferencing to handle non-preferential instances without using the `@` operator. So actually I do want to create another copy of a ring, but one that explicitly depends on an absolute value, so that I can sensibly assign instances that depend on absolute values to it. I think I'm arguing that the `NormedField` and `Valued` classes provide the "bundling" mechanisms, and the `Valued.mk'` and `WithAbs` are two different ways of making that bundling mechanism non-preferential. There is at least some precedence for doing this with type synonyms in [WithLp](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Lp/WithLp.html#The-WithLp-type-synonym), which does something very similar for the $L^p$ norm -- maybe it would be helpful to separate `WithAbs` into a more developed API like this?

smmercuri commented:
Actually, this [`WithIdeal`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Nonarchimedean/AdicTopology.html#WithIdeal) class looks very relevant -- maybe the "structure" approach would indeed work if we have it as a class.

smmercuri commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+derived downstream for `InfinitePlace` (which is a subtype of `AbsoluteValue`).
+
+## Main definitions
+ - `WithAbs` : type synonym equipping a semiring with an absolute value.
```
Thanks! I agree the "equipping" here does not really capture the idea behind using the type synonym. I have expanded here. 

I think I'm not calling it `WithAbsRing` (or I guess that should really be `WithAbsSemiring`) because `Semiring` is required by `AbsoluteValue` and since this synonym aliases the lowest required class for this dependent synonym it seems unambiguous to me -- also see [`WithIdeal`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Nonarchimedean/AdicTopology.html#WithIdeal) which is not called `WithIdealRing` or `WithIdealCommRing` (although this is technically a class rather than a synonym, so I'm not sure if the same naming convention applies). If we were aliasing a `Field` or something else I think I would agree with you on the naming.

smmercuri commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,339 @@
+This file contains the completion of a number field at an infinite place.
+
+This is ultimately achieved by applying the `UniformSpace.Completion` functor, however each
+infinite place defines its own `UniformSpace` instance, so the inference system cannot
```
I think it's an instance -- `UniformSpace` is a class and through `v.1.normedField` we obtain a `UniformSpace` instance (I imagine using `PseudoMetricSpace.toUniformSpace`).

smmercuri commented:
> Actually, this [`WithIdeal`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Nonarchimedean/AdicTopology.html#WithIdeal) class looks very relevant -- maybe the "structure" approach would indeed work if we have it as a class.

Have played around with this idea, and something like this would work:
```
class WithAbs (R S : Type*) [Semiring R] [OrderedSemiring S] where 
  v : AbsoluteValue R S

variable {K : Type*} [Field K] [WithAbs K ℝ]

instance normedField : NormedField K :=
  v.toNormedField

abbrev AbsoluteValue.completion := UniformSpace.Completion K
```
I do later run into inference snags when passing from `AbsoluteValue` to `InfinitePlace`, and then something like this is required:
```
/-- The completion of a number field at an infinite place. -/
abbrev InfinitePlace.completion := letI := WithAbs.mk v.1; AbsoluteValue.completion K
```
For some reason doing `instance : WithAbs K ℝ := WithAbs.mk v.1` before this does not work (probably because it depends on `v`?). I did not go further, but I think from this point on I will have to do a bunch of sign-posting for the inference system to know what's going on. 

Still, I think this would work as a third option. It's maybe less smooth with respect to inferencing than the type synonym approach, but if there are benefits (e.g., type leakage) then we should discuss whether this approach would be better.

faenuccio commented:
Great, thank you so much! I will not have time to dig into this today, but you raised many questions and all of them are interesting, so I will take a little bit to think about all of them and answer properly. It should not take too long.

faenuccio commented:
@erdOne @smmercuri I know I am very late with the review of this PR, and I am very sorry. I will be out of office the next week, but it is my top-1 priority as soon as I will be back, so I hope to provide useful feedback before the end of Nov. Please accept my apologies.

kbuzzard commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,306 @@
+an absolute value on a semiring and returns the semiring. We use this to assign and infer instances
+on a semiring that depend on absolute values. -/
+@[nolint unusedArguments]
+def WithAbs {R S : Type*} [Semiring R] [OrderedSemiring S] :
+    AbsoluteValue R S → Type _ := fun _ => R
```
If we go for this then I don't think it should be in a file called `NumberTheory.NumberField.Completion`, it should probably exist in a new file or at least somewhere which isn't about number fields.

I'm not convinced about the need for this though; I understand the argument, but I'm wondering whether you can just use `abbrev completion := @UniformSpace.Completion K <| letI : NormedField K := v.toNormedField; inferInstance` (and toss in a few more `  letI : NormedField K := v.toNormedField;
` when necessary) to get around this rather than opening up the debate about adding a new definition. Maybe worth having a discussion on Zulip if you can't see my suggestion working. I can see the argument for a new declaration but some people might want to see more than one use case before agreeing.

smmercuri replied:
I agree that it should probably be in its own file, as was done for [`WithLp`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Lp/WithLp.html).

kbuzzard replied:
You could move this declaration to a shortish file WithAbs.lean, put it somewhere appropriate, make some basic API for it e.g. it's a normed ring, it's a normed field if K is a field etc -- just get it out of this file and then hopefully this PR will be ready to be merged. You've convinced me that this approach is worth trying.

smmercuri replied:
Have now added `WithAbs.lean` with a small API modelled on `WithLp.lean`

kbuzzard reviewed (commented):
I've not read through the main file yet but I've reviewed the other changes and left a comment about `WithAbs`. I need the full adele ring for FLT so am actively keen to help getting Salvatore's PRs over the line!

riccardobrasca commented:
Sorry to keep asking changes, but can you split the PR (especially the `WithAbs` part)? Feel free to ask our review. (Hopefully this can be merged quickly since we already discussed it.)

smmercuri commented:
> Sorry to keep asking changes, but can you split the PR (especially the `WithAbs` part)? Feel free to ask our review. (Hopefully this can be merged quickly since we already discussed it.)

Sure, no problem -- I was thinking also that this one has grown quite big

riccardobrasca commented on Mathlib/NumberTheory/NumberField/Embeddings.lean:
```diff
@@ -417,6 +417,19 @@ lemma isComplex_mk_iff {φ : K →+* ℂ} :
+theorem abs_eq_comp (v : InfinitePlace K) :
+    v.1 = Complex.abs.comp v.embedding.injective := by
+  rw [← v.2.choose_spec]
+  rfl
```
Is it possible to replace the `rfl` here by a `rw` or something?

riccardobrasca commented on Mathlib/NumberTheory/NumberField/Completion.lean:
```diff
@@ -0,0 +1,288 @@
+    LocallyCompactSpace (v.completion) :=
+  (isClosedEmbedding_extensionEmbedding_of_comp h).locallyCompactSpace
+
+end AbsoluteValue.Completion
```
Everything until here is unrelated to number fields, right? Is there a better place?

smmercuri replied:
Yes there probably is. They could go in the `WithAbs.lean` file maybe? Or in [`UniformSpace.AbsoluteValue`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/AbsoluteValue.html)? I think `WithAbs.lean` might be a better choice, but let me know if you think otherwise.

riccardobrasca replied:
It seems reasonable also to me. (If you want to open a separate PR we will not complain...)

riccardobrasca commented:
What about a file `.../Completion/Infinite.lean` or something like that? I guess we will want stuff about the completion ad finite places too.

smmercuri commented:
> What about a file `.../Completion/Infinite.lean` or something like that? I guess we will want stuff about the completion ad finite places too.

Completions at finite places are handled [elsewhere](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/DedekindDomain/AdicValuation.html#IsDedekindDomain.HeightOneSpectrum.adicCompletion) at higher generality for fields of fractions of a Dedekind domain, so I'm not sure we will have a finite section here.

riccardobrasca commented:
> > What about a file `.../Completion/Infinite.lean` or something like that? I guess we will want stuff about the completion ad finite places too.
> 
> Completions at finite places are handled [elsewhere](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/DedekindDomain/AdicValuation.html#IsDedekindDomain.HeightOneSpectrum.adicCompletion) at higher generality for fields of fractions of a Dedekind domain, so I'm not sure we will have a finite section here.

OK, no problem. But then maybe `CompletionInfinite.lean` is a better name. (Anyway changing the name of the file is very easy, so this is not important.)

xroblot reviewed (approved):
LGTM. 

Thanks!
